File: internal/lambda/strategies/logging.go
package strategies

import (
	"context"
	"log"
	"os"
	"strings"

	"github.com/coachpo/meltica/internal/schema"
)

// Logging logs all events - useful for debugging.
type Logging struct {
	Logger       *log.Logger
	LoggerPrefix string
}

var loggingSubscribedEvents = []schema.CanonicalType{
	schema.CanonicalType("TRADE"),
	schema.CanonicalType("TICKER"),
	schema.CanonicalType("ORDERBOOK.SNAPSHOT"),
	schema.CanonicalType("EXECUTION.REPORT"),
	schema.CanonicalTypeAccountBalance,
}

// SubscribedEvents returns the list of event types this strategy subscribes to.
func (s *Logging) SubscribedEvents() []schema.CanonicalType {
	return append([]schema.CanonicalType(nil), loggingSubscribedEvents...)
}

func (s *Logging) logger() *log.Logger {
	if s.Logger != nil {
		return s.Logger
	}
	prefix := strings.TrimSpace(s.LoggerPrefix)
	if prefix == "" {
		prefix = "[Logging] "
	}
	s.Logger = log.New(os.Stdout, prefix, log.LstdFlags|log.Lmicroseconds)
	return s.Logger
}

// OnTrade logs trade events.
func (s *Logging) OnTrade(_ context.Context, evt *schema.Event, _ schema.TradePayload, price float64) {
	s.logger().Printf("Trade received: provider=%s symbol=%s price=%.2f", evt.Provider, evt.Symbol, price)
}

// OnTicker logs ticker events.
func (s *Logging) OnTicker(_ context.Context, evt *schema.Event, payload schema.TickerPayload) {
	s.logger().Printf("Ticker: provider=%s symbol=%s last=%s bid=%s ask=%s",
		evt.Provider, evt.Symbol, payload.LastPrice, payload.BidPrice, payload.AskPrice)
}

// OnBookSnapshot logs book snapshot events with truncated orderbook (top 5 levels).
func (s *Logging) OnBookSnapshot(_ context.Context, evt *schema.Event, payload schema.BookSnapshotPayload) {
	logger := s.logger()
	logger.Printf("Book snapshot: provider=%s symbol=%s %d bids, %d asks", evt.Provider, evt.Symbol, len(payload.Bids), len(payload.Asks))

	// Print top 5 bids
	bidLimit := 5
	if len(payload.Bids) < bidLimit {
		bidLimit = len(payload.Bids)
	}
	for i := 0; i < bidLimit; i++ {
		logger.Printf("  BID[%d]: %s @ %s", i, payload.Bids[i].Quantity, payload.Bids[i].Price)
	}

	// Print top 5 asks
	askLimit := 5
	if len(payload.Asks) < askLimit {
		askLimit = len(payload.Asks)
	}
	for i := 0; i < askLimit; i++ {
		logger.Printf("  ASK[%d]: %s @ %s", i, payload.Asks[i].Quantity, payload.Asks[i].Price)
	}
}

// OnOrderFilled logs filled order events.
func (s *Logging) OnOrderFilled(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload) {
	s.logger().Printf("Order filled: provider=%s symbol=%s id=%s qty=%s price=%s",
		evt.Provider, evt.Symbol, payload.ClientOrderID, payload.FilledQuantity, payload.AvgFillPrice)
}

// OnOrderRejected logs rejected order events.
func (s *Logging) OnOrderRejected(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload, reason string) {
	s.logger().Printf("Order rejected: provider=%s symbol=%s id=%s reason=%s", evt.Provider, evt.Symbol, payload.ClientOrderID, reason)
}

// OnOrderPartialFill logs partial fill events.
func (s *Logging) OnOrderPartialFill(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload) {
	s.logger().Printf("Order partial fill: provider=%s symbol=%s id=%s filled=%s remaining=%s",
		evt.Provider, evt.Symbol, payload.ClientOrderID, payload.FilledQuantity, payload.RemainingQty)
}

// OnOrderCancelled logs cancelled order events.
func (s *Logging) OnOrderCancelled(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload) {
	s.logger().Printf("Order cancelled: provider=%s symbol=%s id=%s", evt.Provider, evt.Symbol, payload.ClientOrderID)
}

// OnOrderAcknowledged logs order acknowledgment events.
func (s *Logging) OnOrderAcknowledged(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload) {
	s.logger().Printf("Order acknowledged: provider=%s symbol=%s id=%s", evt.Provider, evt.Symbol, payload.ClientOrderID)
}

// OnOrderExpired logs expired order events.
func (s *Logging) OnOrderExpired(_ context.Context, evt *schema.Event, payload schema.ExecReportPayload) {
	s.logger().Printf("Order expired: provider=%s symbol=%s id=%s", evt.Provider, evt.Symbol, payload.ClientOrderID)
}

// OnKlineSummary logs kline summary events.
func (s *Logging) OnKlineSummary(_ context.Context, evt *schema.Event, payload schema.KlineSummaryPayload) {
	s.logger().Printf("Kline: provider=%s symbol=%s open=%s close=%s high=%s low=%s vol=%s",
		evt.Provider, evt.Symbol, payload.OpenPrice, payload.ClosePrice, payload.HighPrice, payload.LowPrice, payload.Volume)
}

// OnInstrumentUpdate logs instrument catalogue refresh events.
func (s *Logging) OnInstrumentUpdate(_ context.Context, evt *schema.Event, payload schema.InstrumentUpdatePayload) {
	s.logger().Printf("Instrument updated: provider=%s symbol=%s", evt.Provider, payload.Instrument.Symbol)
}

// OnBalanceUpdate logs account balance updates.
func (s *Logging) OnBalanceUpdate(_ context.Context, evt *schema.Event, payload schema.BalanceUpdatePayload) {
	s.logger().Printf("Balance update: provider=%s currency=%s total=%s available=%s",
		evt.Provider, payload.Currency, payload.Total, payload.Available)
}

File: strategies/logging/1.0.0/logging.js
module.exports = {
  metadata: {
    name: "logging",
    tag: "1.0.0",
    displayName: "Logging",
    description: "Emits detailed logs for all inbound events.",
    config: [
      {
        name: "logger_prefix",
        type: "string",
        description: "Prefix prepended to each log message",
        default: "[Logging] ",
        required: false
      },
      {
        name: "dry_run",
        type: "bool",
        description: "When true, strategy logs intended orders without submitting them",
        default: true,
        required: false
      }
    ],
    events: [
      "Trade",
      "Ticker",
      "BookSnapshot",
      "ExecReport",
      "KlineSummary",
      "BalanceUpdate",
      "RiskControl",
      "InstrumentUpdate"
    ]
  },
  create: function (env) {
    const prefixSource = (env.config.logger_prefix ?? "[Logging] ").trim();
    const prefix = prefixSource.length > 0 ? prefixSource : "[Logging] ";
    const log = (...parts) => env.helpers.log(`${prefix}${parts.join(" ")}`);

    function formatPrice(price) {
      if (price === undefined || price === null) {
        return "";
      }
      if (typeof price === "number" && Number.isFinite(price)) {
        return price.toFixed(2);
      }
      return String(price);
    }

    function summarizeOrders(orders, side) {
      if (!Array.isArray(orders) || orders.length === 0) {
        return;
      }
      const limit = Math.min(orders.length, 5);
      for (let i = 0; i < limit; i += 1) {
        const level = orders[i] || {};
        log(`${side}[${i}]:`, `${level.quantity ?? level.qty ?? ""} @ ${level.price ?? ""}`);
      }
    }

    return {
      wantsCrossProviderEvents: function () {
        return true;
      },
      onTrade: function (_ctx, evt, _payload, price) {
        log("========== JavaScript version ==========");
        log(
          "Trade received:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `price=${formatPrice(price)}`
        );
      },
      onTicker: function (_ctx, evt, payload) {
        log("========== JavaScript version ==========");
        log(
          "Ticker:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `last=${payload?.lastPrice ?? ""}`,
          `bid=${payload?.bidPrice ?? ""}`,
          `ask=${payload?.askPrice ?? ""}`
        );
      },
      onBookSnapshot: function (_ctx, evt, payload) {
        log("========== JavaScript version ==========");
        const bids = payload?.bids ?? payload?.Bids ?? [];
        const asks = payload?.asks ?? payload?.Asks ?? [];
        log(
          "Book snapshot:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `${bids.length} bids, ${asks.length} asks`
        );
        summarizeOrders(bids, "BID");
        summarizeOrders(asks, "ASK");
      },
      onOrderFilled: function (_ctx, evt, payload) {
        log(
          "Order filled:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`,
          `qty=${payload?.filledQuantity ?? ""}`,
          `price=${payload?.avgFillPrice ?? ""}`
        );
      },
      onOrderRejected: function (_ctx, evt, payload, reason) {
        log(
          "Order rejected:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`,
          `reason=${reason ?? payload?.rejectReason ?? ""}`
        );
      },
      onOrderPartialFill: function (_ctx, evt, payload) {
        log(
          "Order partial fill:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`,
          `filled=${payload?.filledQuantity ?? ""}`,
          `remaining=${payload?.remainingQty ?? ""}`
        );
      },
      onOrderCancelled: function (_ctx, evt, payload) {
        log(
          "Order cancelled:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`
        );
      },
      onOrderAcknowledged: function (_ctx, evt, payload) {
        log(
          "Order acknowledged:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`
        );
      },
      onOrderExpired: function (_ctx, evt, payload) {
        log(
          "Order expired:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `id=${payload?.clientOrderId ?? ""}`
        );
      },
      onKlineSummary: function (_ctx, evt, payload) {
        log(
          "Kline:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${evt?.symbol ?? ""}`,
          `open=${payload?.openPrice ?? ""}`,
          `close=${payload?.closePrice ?? ""}`,
          `high=${payload?.highPrice ?? ""}`,
          `low=${payload?.lowPrice ?? ""}`,
          `vol=${payload?.volume ?? ""}`
        );
      },
      onInstrumentUpdate: function (_ctx, evt, payload) {
        const instrument = payload?.instrument ?? payload?.Instrument ?? {};
        log(
          "Instrument updated:",
          `provider=${evt?.provider ?? ""}`,
          `symbol=${instrument.symbol ?? instrument.Symbol ?? ""}`
        );
      },
      onBalanceUpdate: function (_ctx, evt, payload) {
        log(
          "Balance update:",
          `provider=${evt?.provider ?? ""}`,
          `currency=${payload?.currency ?? ""}`,
          `total=${payload?.total ?? ""}`,
          `available=${payload?.available ?? ""}`
        );
      },
      onRiskControl: function (_ctx, _evt, payload) {
        log(
          "Risk control:",
          `strategy=${payload?.strategyID ?? payload?.strategyId ?? ""}`,
          `status=${payload?.status ?? ""}`,
          `breach=${payload?.breachType ?? payload?.breach_type ?? ""}`,
          `reason=${payload?.reason ?? ""}`,
          `metrics=${JSON.stringify(payload?.metrics ?? {})}`,
          `killSwitch=${payload?.killSwitchEngaged ?? payload?.kill_switch_engaged ?? false}`,
          `circuitBreaker=${payload?.circuitBreakerOpen ?? payload?.circuit_breaker_open ?? false}`
        );
      }
    };
  }
};
